### ISSUE TITLE
Enable atomic `send_batch` for SourceSender

### ISSUE THREAD

**Issue by lukesteensen** [vectordotdev/vector#10975]
For the purpose of propagating backpressure, it would be helpful to have a method on `SourceSender` that atomically accepted or rejected a batch of events from a source. This would prevent the situation where a source receives a number of events, begins to send them as a batch, but becomes blocked when the inner channel fills up after having accepted some but not all of the events.

The `send_batch` method was recently added to `SourceSender` with the intent of filling this gap, but it currently is only a partial solution. It's limited by the fact that communication between the sender and the pump task is still one event at a time. This means it's possible to become blocked in a state where some events from a batch have been accepted upstream and others haven't. 

There are two high-level strategies for resolving this:

1. Change the communication between the sender and pump task to pass `Vec<Event>` or similar. The challenge here becomes how to size the channel. Because some source still send an event at a time, even under load, there needs to be a sufficiently large buffer to enable smooth operation. But with a large channel buffer, the maximum number of events buffered for sources that do send large batches at a time would end up very large, potentially ballooning memory requirements and making backpressure propagation less responsive.
2. Move forward with #10912 and rely on the broader process of #9961 to take care of the issue at the fanout directly.

**Comment by lukesteensen** [comment]
@bruceg it seems to me that this is _kind of_ done with the recent array work, with the caveat that internally it can technically be up to one send operation per event type. In practice, that should not be a problem for most sources. Does that sounds right to you?

**Comment by bruceg** [comment]
Yes, the scope of the problem will definitely have been reduced with #11587. It has not been eliminated, as some sources (`vector` `datadog_agent`?) could send batches with multiple event types, resulting in multiple arrays sent to the output. With my most recent change, even large batches could cause multiple arrays being output. So yes, it is _kind of_ done, _probably_ not a problem for most sources, but it's a still a potential issue if we are concerned.

**Comment by jszwedko** [comment]
We have also discussed requiring different event types to be different component outputs https://github.com/vectordotdev/vector/issues/10248
